### ISSUE TITLE
EHN: inherited configuration files.

### ISSUE THREAD

**Issue by musically-ut** [jwilm/alacritty#779]
I use alacritty on multiple platforms such that the dotfiles between them is kept in sync (using softlinks and Dropbox, though that is not important).

However, I would like to have slightly tweaked configuration on each system (e.g. larger font sizes on one system, `use_thin_strokes` on one system, light/dark color themes, etc.).

Right now, the config is read only from one single file:

https://github.com/jwilm/alacritty/blob/8990aa4568ea8a7481f827512b7487be4d39332a/src/main.rs#L65-L75

Would there be interest in converting `config_path` to an array and such that all the configs files are read in the said order:

    /// 1. $XDG_CONFIG_HOME/alacritty/alacritty.yml
    /// 2. $XDG_CONFIG_HOME/alacritty.yml
    /// 3. $HOME/.config/alacritty/alacritty.yml
    /// 4. $HOME/.alacritty.yml

and the configurations read from each is _merged_ with preferences being overwritten by files read later.

This would allow me to keep, say, `$HOME/.config/alacritty/alacritty.yml` synced between the systems while keeping local configuration in `$HOME/.alacritty.yml`.

Alternatively, we can add a file to the list, i.e. `$HOME/.alacritty.local.yml`, which will be exclusively used to override the preferences read from other files.

**Comment by lukaslueg** [comment]
Related to #169

**Comment by mathstuf** [comment]
I'd prefer there be another filename that is used rather than using the same one in different locations to be the trigger. One of the benefits of XDG paths is the decluttering of `$HOME`.

**Comment by 2m** [comment]
One more advantage of merging config files is the easier management of config as alacritty evolves. This would allow having only the custom settings in a config file, and all of the default ones would be merged in from the default config.

**Comment by nixpulvis** [comment]
Reducing the filesize of `.config/alacritty/alacritty.yml` is a nice win indeed.

Sidenote: it drives me crazy that it's named `alacritty.yml` and not `config` or `config.yml`, even though it hardly matters.

**Comment by mathstuf** [comment]
Alternatively, a directory could be read instead of specific paths. For example, anything matching `$XDG_CONFIG_HOME/alacritty/conf.d/*.yml` is read in ASCII-betical order and later files override previous files.

**Comment by 2m** [comment]
I thought about merging multiple yml files as well, but then there would still be no way to override mouse and keyboard bindings without modifying the original configuration file. Here is an example mouse binding:
```yml
mouse_bindings:
  - { mouse: Middle, action: PasteSelection }
```
and keyboard
```yml
key_bindings:
  - { key: V,        mods: Control|Shift,    action: Paste               }
  - { key: C,        mods: Control|Shift,    action: Copy                }
```
We would either need to write special bindings merging logic that merges entries in the array by `mouse` and `key` + `mods` attributes, or refactor the config files and have these attributes as keys in the yml.

**Comment by mathstuf** [comment]
Why would it need to be special logic? Or do you mean "more complicated than `HashMap::extend`" when loading configuration files? The keybindings is a list. I'd just have it append all of those entries together and then `.collect()` them into the final keybinding table. There may need to be a special way to say "bind this key to no action" to "delete" a prior binding.

**Comment by jokeyrhyme** [comment]
I wonder if we can merge the default YAML files a little bit more? What happens if we have both Super and Command keybindings for a given OS?

**Comment by mathstuf** [comment]
I'd expect that you'd have multiple keybinds which do the same action. This would be useful for Apple hardware with non-Apple keyboards, but I guess macOS does its own translations internally?

**Comment by jedahan** [comment]
For this to work, every piece of software that is cross-platform, but with different settings per platform, would have to implement their own logic for merging and overwriting configs. I would much rather have this logic exist in whatever syncing/dotfile management tool is being used, than have more complicated inheritance in multiple apps.

I also sync my dotfiles across platforms using plain old git ( http://github.com/jedahan/dotfiles ), and just have two branches, on for macOS and one for linux. I could imagine some sort of cat + diff + cut + paste magic, or something more robust as post receive hooks to merge files automatically...

**Comment by nixpulvis** [comment]
TLDR: Keep configs DRY, allow people to use one config for macOS and Linux.

I'm not sure we *need* to allow multiple config files and a inheritance (with some precedence rules). What I *do* think we should do is give each config option a default (hardcoded) so that a config file is optional, and then make the config file simply overrule the defaults. This will make a user's config file only related to things he/she cares about, which in turn solves, at least my own cross platform issue, since the main settings I care about are colors and fonts.

I do however think a more general approach would be cool since it doesn't make life harder for the user, but simply allows for more kinds of configurations without external tools.

**Comment by arooni** [comment]
So what's the solution right now?  I have different fonts I like to look at on mac and ubuntu/linux enviornments.  what can be done?

**Comment by NickeZ** [comment]
How would you disable built in keybindings in the case where we merge the config file with the default settings?

It would also be cool if you could override any config from the command line.

**Comment by chrisduerr** [comment]
@NickeZ There's #1258 to aim exploring the expansion of command line parameters.

I'm also personally not a big fan of the idea to merge multiple configuration files together. I feel like the complexity here shouldn't be introduced to Alacritty, but to an external system instead.

Having different configuration files for different systems is certainly not something new to Alacritty. On Linux there's about a billion dotfile managers aimed at helping out with this, I personally think that just using git for that is a great solution. Using separate branches for separate systems or just using the stash for system-specific settings has worked extremely well for me.

**Comment by nixpulvis** [comment]
@chrisduerr I agree with you in general that we shouln't be overly concerned with solving all the problems of user configuration envitronments. There are other better tools... That said, I'm strongly in favor of something that would allow us to unify the macOS/Linux/Window configs, and only provide the differing subsets as OS specific files. I think that a sane solution for this could help with some peoples issues as well.

**Comment by chrisduerr** [comment]
@nixpulvis Oh yeah, my comments were just about merging multiple different configuration files together. I'm 100% in favor of merging the macos/linux/windows configuration files together, that will actually save us a lot of work in the long run.

I don't think it's necessary to provide different files for that though, things like default fonts could be handled by making the value optional for example and then substituting it in the code at runtime when it is not present in the config. That allows specifying default system values at compile-time.

**Comment by nixpulvis** [comment]
@chrisduerr I'm more thinking about the keybindings... I don't see a good way to deal with them without some kind of `include` or conditional `include` like feature. "Inherited" configuration to me is just an implicit `include`.

**Comment by jokeyrhyme** [comment]
> 
> 
> @chrisduerr I'm more thinking about the keybindings... I don't see a good way to deal with them without some kind of `include` or conditional `include` like feature. "Inherited" configuration to me is just a system of implicit `include`s.

What if everything that was bound to Super + blah was also bound to Ctrl + Shift + blah (and vice versa) by default, even when there is no config file, but we retained the ability to remap / unbind via the config file as we like

There really aren't that many bindings once we collapse the macOS + Linux + Windows ones together, as we've had to avoid common shell and window manager keybindings anyway (and really, a terminal emulator doesn't have that many feature to bind)

**Comment by jokeyrhyme** [comment]
> 
> 
> @nixpulvis I don't think it's necessary to provide different files for that though, things like default fonts could be handled by making the value optional for example and then substituting it in the code at runtime when it is not present in the config. That allows specifying default system values at compile-time.

Fonts are a special case, in that we could support listing multiple fonts and falling back when they are not found, like in CSS, VSCode, MacVim, etc

But yes, I'm all for compile-time defaults: ideally I want no config file at all, or a very minimal one for personal tweaks

**Comment by nixpulvis** [comment]
@jokeyrhyme well we just got our wish, #1993 just landed on `master`. :tada:

**Comment by norcalli** [comment]
I did it in #2113, but I have made some small design decisions which you can see there.

**Comment by bugeats** [comment]
The vim community has addressed this problem using a mix of feature-detection and explicit os branches in config scripts. Example:

```vim
if has("gui_running")
    if g:os == "Darwin"
        set guifont=Fira\ Mono:h12
    elseif g:os == "Linux"
        set guifont=Fira\ Mono\ 10
    elseif g:os == "Windows"
        set guifont=Fira_Mono:h12:cANSI
    endif
endif
```

I'm not sure what the most elegant yaml equivalent would be, but I'm sure it's possible.

I would also vote for strictly using **feature-detection** instead of OS name flags. Feature detection is how many JS projects survived the browser dark ages and remain maintainable today.

**Comment by chrisduerr** [comment]
> Feature detection is how many JS projects survived the browser dark ages and remain maintainable today.

Alacritty is not actually a JavaScript project or has anything to do with Browsers at all. Thankfully.

There's absolutely no way conditional logic like that will ever be added to Alacritty's configuration file. It's a configuration file, not a programming language.

**Comment by bugeats** [comment]
@chrisduerr I am quite aware of what kind of project Alacritty is and what language it's written in. What I am trying to do is provides a examples of successful strategies for running the same application in disparate environments in a future-proof way. This is most exemplified in how the web has evolved.

For example, my mac has a high DPI screen, and I would like to use a different font in that case. It's more accurate to say that I want a certain config in the presence of actual system features rather than the platform at large. 

As for conditional logic, it would not need to be any more complicated than a set of config values prefixed by a namespace. Keys would be merged based on feature presence.

For example of an elegant language that that was built for both configuration and conditional expression, see [cuelang](https://cuelang.org/). Thankfully there are people out there working on this problem that have more imagination than you just demonstrated.

**Comment by chrisduerr** [comment]
> What I am trying to do is provide examples of successful strategies for running the same application in disparate environments in a future-proof way.

That already exists, you can easily achieve this using `git`. Alacritty should stick to what it's good at and leave issues like these to applications that focus on handling that problem.

> As for conditional logic, it would not need to be any more complicated than a set of config values prefixed by a namespace. Keys would be merged based on feature presence.

That's already way too complicated. Alacritty should strive to simplify its configuration file, not make it more complicated. It's already stretching the bounds of what is bearable.

> For an example of an elegant language that that was built for both configuration and conditional expression, see cuelang.

There is zero desire for a language that supports configuration **and conditional expression**, as stated in my previous comment.

**Comment by 2m** [comment]
I would like to chip-in my two cents as someone who saw a need for advanced configuration for alacritty, but does not anymore.

I have solved different-computers-need-different-config usecase for alacritty and all of the other tools I use with [chezmoi](https://github.com/twpayne/chezmoi).

Here is an example of my alacritty config template file, that sets the font size according to the machine with default being `10`:

https://github.com/2m/dotfiles/blob/master/home/dot_config/alacritty/alacritty.yml.tmpl#L134-L136

And then chezmoi configuration allows to keep overrides for different machines. On my MacBook with retina display I use font size of `13`:

https://github.com/2m/dotfiles/blob/master/chezmoi-home.toml#L45

Since chezmoi was created to be doing exactly that: *Manage your dotfiles across multiple machines* it integrates really well with git, has support for templates, allows getting secrets from secure vaults, etc.

So I would like to encourage you to try chezmoi and focus alacritty effort to do what it does, being terminal emulator, the best.

**Comment by thomasf** [comment]
After getting very tired of having to update almost identical alacritty.yml's for each of my installation I quickly wrote a very small utility to merge multiple files.

Basically it just merges these files together common + colors + detected os + detected hostname

```
c.common.yml
c.colors.solarized-light.yml
c.os.darwin.yml
c.os.linux.yml
c.host.crangy.yml
c.host.fogskum.yml
c.host.prizza.yml
c.host.transwhale.yml
```

It's written in go and is very basic but it seems to fit my needs
https://github.com/thomasf/dotfiles-thomasf-xmonad/tree/master/.config-xmonad/alacritty

You can also override default/detected options with command line flags.

```
$ go run . -h
Usage of /tmp/go-build462102881/b001/exe/alacritty-genconfig:
  -host string
    	host name (default "transwhale")
  -os string
    	os name (default "linux")
  -out string
    	output filename (default "alacritty.yml")
  -theme string
    	theme name (default "solarized-light")
```

if anyone is interested in using this I believe that the yaml parser used by alacritty is faulty because using values like `y`, `Y`, `n` and `N` converts them to boolean with this parser so they have to be quoted. It seems like this is how several yaml parsers works and maybe it is the correct way to parse, idk., yaml is kind of weird.

```
 - { key: 'Y',                           mode: Vi, action: Copy                    }
...
font:
  size: 16.0
  offset:
    x: 0
    'y': 1

```
